### PR TITLE
Split out a `Union` type from `Variant`

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit splits out the union specialization from `variant`,
finalizing the transition to the new AST of types available in the
current component model draft. This is mostly like the prior
splittings-out where there's some duplication but it's relatively
localized and hopefully not too hard to follow.

### CODE REVIEW

File: crates/parser/src/abi.rs [hunk 1]
```diff
@@ -568,6 +568,21 @@ def_instruction! {
             ty: TypeId,
         } : [1] => [1],
 
+        /// TODO
```
Comment (peterhuene) [inline crates/parser/src/abi.rs:571]: TODO comment here and below.

**Review by peterhuene — APPROVED** [review]
Looks good 👍 Just the `TODO` comments, I think.

File: crates/parser/src/abi.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/parser/src/abi.rs:571]: Gah sorry I swear I'm not forgetting these intentionally, despite this being present in every PR I've sent so far...
